feat: add `facadecheck` script (#17769)
juju/juju: doc/architectural-overview.md

## API

Juju controllers expose an API endpoint over a websocket connection. Related
API endpoints are grouped into a set called a "facade". Each facade generally
serves either the client, the controller, or Juju agents. As a general rule,
each worker that needs to talk to the API server will get its own facade (for
example, `Provisioner`, `Upgrader`, and `Uniter`, each used by the eponymous
worker types).

The API server is implemented in the `apiserver` top level package. Facades
live inside the `apiserver/facades` package. There is a subpackage `agent` for
agent facades, `client` for client facades, and `controller` for controller
facades.

Various facades share functionality; for example, the Life method is used by
many worker facades. In these cases, the method is implemented on a separate

> type in the `apiserver/common` package, then embedded in the facade.

Reviewer: > the method is implemented on a separate type in the `apiserver/common` package, then embedded in the facade.

This is an anti-pattern. Anything you expose via embedding is then made public and needs to be explicitly versioned. What you should do, which is important for the document and to document it, is to make the common part a field and expose what you need. Thus, allowing you to version on common methods.

Pseudo code for this would be:

```go
type Facade struct {
     lifer common.Life
}

func (f Facade) Life(ctx context.Context, modelArgs params.LifeArgs) (params.LifeResult, error) {
     return f.lifer.Life(ctx, modelArgs)
}
```

Again, this allows us to move away from Life. You can't do that without breaking all facades that embed common APIs.

I would also note that any public functions on the facade become part of the facade's API.

Author: I agree with you, I'm just describing what I've seen in many facades.

Each facade is separately versioned. For any Juju version, you can see the list
of supported versions for each facade in `api/facadeversions.go`. When using
two different versions of Juju, they will "negotiate" to find a version of the
facade which is supported by both. A facade's version must be incremented when
the interface changes, and old facade versions can generally only be removed in
a new major version. You can use the `facadecheck` script in the `scripts`
folder to compare supported facade versions for any two versions of Juju.

> All APIs *should* be implemented such that they can be called in bulk. Newer
> facades generally follow this rule, but older ones may not.

Reviewer: ```suggestion
Moving forward all APIs *should* be implemented such that they can be called in bulk.
```
